### PR TITLE
Add Magento 2 NVP Method

### PR DESCRIPTION
**Pull request by tonyisworking**
- Rename to payment-paypal-magento2
- Refactored for Magento2 and compatible with 2.2.x built in Paypal Express - able to read transaction data from Paypal and do Refunds
- Use NVP methods similar to how Magento 2 implements paypal
- Update to send cart items, addresses, totals, discount, shipping to Paypal
- Disable Address changes in Paypal, use the checkout address
- Add ability to use Standard Rest implementation
- Add `setExpressCheckout` endpoint for api

Tested on US Regions 1.10.x and Magento 2.2.x

### CODE REVIEW

File: hooks/beforeRegistration.ts [hunk 1]
```diff
@@ -1,35 +1,16 @@
...
+  const VSF_PAYPAL_CODE = 'paypal_express'
 
-  if (!Vue.prototype.$isServer) {
+  if (!isServer) {
     const storeView = currentStoreView()
     const { currencyCode } = storeView.i18n
-    const clientId = config.paypal.clientId
-    const sdkUrl = `https://www.paypal.com/sdk/js?client-id=${clientId}&currency=${currencyCode}`
+    const clientId = config.paymentPaypalMagento2.clientId
+    const sdkUrl = `https://www.paypal.com/sdk/js?client-id=${clientId}&currency=${currencyCode}&disable-funding=card,credit&intent=authorize`
```
Comment (dimasch) [inline hooks/beforeRegistration.ts]: @tonyisworking Need move the load paypal sdk to the mount level of button compoent.

File: hooks/afterRegistration.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  if (!isServer) {
+    Vue.prototype.$bus.$on('set-unique-payment-methods', methods => {
+      store.commit('payment-paypal-magento2/' + types.SET_BACKEND_PAYMENT_PAYPAL_EXPRESS, methods)
+    })
+
+    Vue.prototype.$bus.$on('checkout-before-placeOrder', placeOrder)
+
+    // Mount the info component when required
+    Vue.prototype.$bus.$on('checkout-payment-method-changed', (paymentMethodCode) => {
+      if (paymentMethodCode === VSF_PAYPAL_CODE) {
```
Comment (dimasch) [inline hooks/afterRegistration.ts]: @tonyisworking We can improve 
`correctPaymentMethod = paymentMethodCode === VSF_PAYPAL_CODE`

File: hooks/beforeRegistration.ts [hunk 3]
```diff
@@ -1,35 +1,16 @@
...
-  const VSF_PAYPAL_CODE = 'vsfpaypal'
-  store.dispatch('payment/addMethod', {
-    'title': 'Paypal',
-    'code': VSF_PAYPAL_CODE,
-    'cost': 0,
-    'costInclTax': 0,
-    'default': false,
-    'offline': true
-  })
+  // const VSF_PAYPAL_CODE = 'vsfpaypal'
```
Comment (dimasch) [inline hooks/beforeRegistration.ts]: @tonyisworking We can remove the unnecessary comments.
Thanks! 
I will push an improvements to this branch, if you will allow

File: components/Button.vue [hunk 4]
```diff
@@ -1,55 +1,217 @@
...
+        }).then((result) => {
+          this.tokenId = result.token
+          return this.tokenId
+        })
+      })
+    },
+    async createOrderRest (data, actions) {
+      return store.dispatch('cart/syncTotals', {
+        methodsData: {
+          country: store.state.checkout.shippingDetails.country,
```
Comment (dimasch) [inline components/Button.vue]: @tonyisworking Dublicated code lines...

File: components/Button.vue [hunk 5]
```diff
@@ -1,55 +1,217 @@
...
+          tax: {
+            currency_code: this.currencyCode,
+            value: ''
+            // optional tax already set in totals, this is not needed
+            // value: (product.totals.price_incl_tax - product.totals.price).toFixed(2)
+          },
+          description: (product.options && product.options.length > 0) ? product.options.map((el) => { return el.value }).join(',') : '',
+          quantity: product.qty,
+          sku: product.sku,
+          category: 'PHYSICAL_GOODS'
```
Comment (dimasch) [inline components/Button.vue]: @tonyisworking Hardcoded :)

File: README.md [hunk 6]
```diff
@@ -1,22 +1,55 @@
...
+$ git clone git@github.com:tonyisworking/vsf-payment-paypal-magento2.git ./vue-storefront/src/modules/payment-paypal-magento2
 ```
 
 ```json
-"paypal": {
+"paymentPaypalMagento2": {
   "clientId": "",
   "endpoint": {
-    "complete": "http://localhost:8080/api/ext/paypal/complete"
+    "complete": "http://localhost:8080/api/ext/paypal/complete",
```
Comment (dimasch) [inline README.md]: @tonyisworking For the `default.json` need use:
`"complete": "/api/ext/paypal/complete"` without `http://localhost:8080` - this only for  `local.json`

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,16 +1,29 @@
 # Changelog
+
 All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## 2019.30.10
```
Comment (dimasch) [inline CHANGELOG.md]: @tonyisworking We change the open api intefaces - skip backward compatibility.
Need set:
`## [1.0.0] - 2019.30.10`

File: store/actions.ts [hunk 8]
```diff
@@ -5,8 +5,21 @@ import { adjustMultistoreApiUrl } from '@vue-storefront/core/lib/multistore'
 
 // it's a good practice for all actions to return Promises with effect of their execution
 export const actions: ActionTree<PaypalState, any> = {
-  complete ({}, params) {
-    let url = config.paypal.endpoint.complete
+  complete({ }, params) {
+    let url = config.paymentPaypalMagento2.endpoint.complete
```
Comment (dimasch) [inline store/actions.ts]: @tonyisworking All action should return a `Promise`, for use `async/await` in methods.
+ @collymore

File: components/Button.vue [hunk 9]
```diff
@@ -1,55 +1,217 @@
...
+          zipCode: store.state.checkout.shippingDetails.zipCode,
+          region: store.state.checkout.shippingDetails.region,
+          region_id: store.state.checkout.shippingDetails.regionId,
+          region_code: store.state.checkout.shippingDetails.regionCode,
+          method_code: store.state.checkout.shippingDetails.shippingMethod,
+          carrier_code: store.state.checkout.shippingDetails.shippingCarrier,
+          payment_method: null
+        },
+        forceServerSync: true
+      }).then(() => {
```
Comment (dimasch) [inline components/Button.vue]: @tonyisworking If we use `async` don't need use `then`, now simply use `await`
+ @collymore

File: hooks/beforeRegistration.ts (hunk 3 above)
Comment (tonyisworking) [reply hooks/beforeRegistration.ts]: > I will push an improvements to this branch, if you will allow

@dimasch Yes you can do that.

File: components/Button.vue (hunk 4 above)
Comment (tonyisworking) [reply components/Button.vue]: @dimasch what do you mean?

File: hooks/beforeRegistration.ts (hunk 3 above)
Comment (tonyisworking) [reply hooks/beforeRegistration.ts]: Do i need to anything for that?

**Pull request closed without merging**